File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> result;
    vector<int> path;
    vector<bool> used;

    void backtrack(vector<int>& nums) {
        if (path.size() == nums.size()) {
            result.push_back(path);
            return;
        }

        for (int i = 0; i < nums.size(); i++) {
            if (used[i]) continue;
            used[i] = true;
            path.push_back(nums[i]);
            backtrack(nums);
            path.pop_back();
            used[i] = false;
        }
    }

    vector<vector<int>> permute(vector<int>& nums) {
        used.assign(nums.size(), false);
        backtrack(nums);
        return result;
    }
};

File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []

        def f(curr):
            if len(curr) == len(nums):
                res.append(curr[:])
                return

            for num in nums:
                if num in curr:
                    continue

                curr.append(num)
                f(curr)
                curr.pop()

        f([])
        return res
